src/main/java/com/dj/core/optimizer/SGDOptimizer.java

```diff
@@ -12,39 +13,55 @@
     private final Loss loss;
     private final int numberOfEpochsToTrain;
     private final OptimizerProgressListener lossCalculatedListener;
+    private final boolean dynamicallyAdjustLearningRate;
+    private double currentLoss = Double.MIN_VALUE;
 
     public SGDOptimizer(final Loss loss,
                         final int numberOfEpochsToTrain,
-                        final OptimizerProgressListener lossCalculatedListener) {
+                        final OptimizerProgressListener lossCalculatedListener,
+                        final boolean dynamicallyAdjustLearningRate) {
         this.loss = loss;
         this.numberOfEpochsToTrain = numberOfEpochsToTrain;
         this.lossCalculatedListener = lossCalculatedListener;
+        this.dynamicallyAdjustLearningRate = dynamicallyAdjustLearningRate;
     }
 
     public SGDOptimizer(final Loss loss,
-                        final int numberOfEpochsToTrain) {
-        this(loss, numberOfEpochsToTrain, null);
+                        final int numberOfEpochsToTrain,
+                        final boolean dynamicallyAdjustLearningRate) {
+        this(loss, numberOfEpochsToTrain, null, dynamicallyAdjustLearningRate);
     }
 
     @Override
     public void train(
+            final Context context,
             final List<Neuron> inputNeurons,
             final List<Neuron> outputNeurons,
             final double[][] inputData,
             final double[][] expectedResult,
             final double[][] inputTestData,
             final double[][] expectedTestResult) {
          IntStream.range(0, numberOfEpochsToTrain).forEach(epoch -> {
-             if (lossCalculatedListener != null) {
+             if (lossCalculatedListener != null || dynamicallyAdjustLearningRate) {
                  final double loss = calculateLoss(
                          inputNeurons,
                          outputNeurons,
                          inputTestData,
                          expectedTestResult);
-                 lossCalculatedListener.onProgress(
-                         loss,
-                         epoch,
-                         numberOfEpochsToTrain);
+                 if (lossCalculatedListener != null) {
+                     lossCalculatedListener.onProgress(
+                             loss,
+                             epoch,
+                             numberOfEpochsToTrain);
+                 }
+                 if (dynamicallyAdjustLearningRate && currentLoss < loss) {
+                     System.out.printf("Loss have increased from: %f to %f\n", currentLoss, loss);
+                     System.out.printf("Changing learning rate from: %f to %f\n",
+                             context.getLearningRate(),
+                             context.getLearningRate() / 2.);
+                     context.setLearningRate(context.getLearningRate() / 2.);
```

Comment:
@b0noI I am not sure, but we can store the '2.' value in a context as a divider.

Reply:
I see the benefits of making it adjustable, however, IMHO it should not go to the Context, since context is in charge of the graph, while 2. is something that only the optimizer is cares about. I need to think about it, maybe we can make it a field of the optimizer.

Reply:
Magic number now also remove, WDYT can we merge it?